src/sentry/middleware/ratelimit.py

```diff
@@ -23,8 +22,18 @@
 )
 
 
-class RatelimitMiddleware(MiddlewareMixin):
-    """Middleware that applies a rate limit to every endpoint."""
+class RatelimitMiddleware:
+    """Middleware that applies a rate limit to every endpoint.
+    See: https://docs.djangoproject.com/en/4.0/topics/http/middleware/#writing-your-own-middleware
+    """
+
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request: Request) -> Response:
+        response = self.get_response(request)
+        self.process_response(request, response)
+        return response
```

Comment:
since we're doing this change, I think it would be better to remove the `process_view` function and use the new middleware functionality in django more directly i.e.:

```python3
def __call__(self, request: Request):
    # whatever is done in `process_view` should be done here

    response = self.get_response(request)
    # whatever is done in `process_response` should be done here
```

That doesn't mean that having separate functions is bad but for me there are two reasons why what we have now is not ideal:

 1. Looking at this code, unless you know how django middleware works, it's not clear when `process_view` would ever be called. (I had to look it up myself because I had no idea how the tests weren't failing).
 2. There is some bad behavior we can clean up while we're here. You'll notice that we attach fields to the `request` object in `process_view` with the hope that they'll be there once we hit `process_response`. Example:
 
```python3
request.will_be_rate_limited = False
request.rate_limit_category = None
request.rate_limit_uid = uuid.uuid4().hex
request.rate_limit_key = get_rate_limit_key(view_func, request)
```
If these variables are in the same function scope, we can reduce the amount of baggage the request object needs to have. If we do that, the middleware component can be more self-contained and depend less on the global scope in the `request` object

Reply:
So we tried using the function approach yesterday. One problem is that we use data attached to the `view_func` for determining the rate limit, like the overrides. the `view_func` argument isn't available in the function middleware.

Reply:
^^ I tried again after reading this comment and got stuck w/o `view_func`, `get_rate_limit_key` relies on it. At the bare minimum what if I left a comment in `__call__` that says Django is calling `process_view`?

Reply:
It's up to you whether you want to do this. At the end of the day you're going to own this. In my opinion the code is is is a little clunky, specifically the rate limiter intreracttion

```
def process_view():
    metadata = above_rate_limit_check() # this checks the rate limiter and registers it in the concurrent rate limiter

def process_response():
    finish_request(request.rate_limit_key, request.uid) # this has to be called otherwise the concurrent slot will leak until cleanup is done
 ```
 
There is a more elegant way to do this if the request and response are in the same scope.
 
 Example pseudocode
 ```
 try:
     with rate_limit_scope(rate_limit_value):
        # do the request
 except RateLimitExceeded as e:
    return HTTPResponse(e.rate_limit_data)
```

However at the end of the day it's up to you what tradeoffs you want to make. If the way it is right now is ok with you, then it's ok with me